### PR TITLE
Normalize Regexes during build

### PR DESCRIPTION
**Pull request by ignacy123**

### CODE REVIEW

File: app/src/test/kotlin/compiler/lexer/regex/RegexFactoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        val union1 = RegexFactory.createUnion(CONCAT_AB, CONCAT_AC)
+        val union2 = RegexFactory.createUnion(CONCAT_AC, CONCAT_BC)
+        assertEquals(RegexFactory.createUnion(union1, CONCAT_BC), RegexFactory.createUnion(CONCAT_AB, union2))
+        assertEquals(RegexFactory.createUnion(CONCAT_AB, EMPTY), EMPTY)
+        assertEquals(RegexFactory.createUnion(EMPTY, CONCAT_AB), EMPTY)
+    }
+
+    @Test fun testUnionNormalizeOrdering() {
+        var unionLeftOrdered = RegexFactory.createUnion(CONCAT_AB, CONCAT_AC)
```
[the remark this answers is not in the document]
Comment (ignacy123) [reply app/src/test/kotlin/compiler/lexer/regex/RegexFactoryTest.kt:26]: How about a multiline definition?

>         val unionLeftOrdered = RegexFactory.createUnion(
>            RegexFactory.createUnion(
>                RegexFactory.createUnion(CONCAT_AB, CONCAT_AC), CONCAT_BC
>            ),
>            CONCAT_CD
>        )

[the remark this answers is not in the document]
Comment (JHodorHodor) [reply app/src/test/kotlin/compiler/lexer/regex/RegexFactoryTest.kt:26]: Much better, maybe even better than named variables.

[the remark this answers is not in the document]
Comment (JHodorHodor) [reply app/src/test/kotlin/compiler/lexer/regex/RegexFactoryTest.kt:26]: maybe even deeper, like
```
    val unionLeftOrdered = RegexFactory.createUnion(
       RegexFactory.createUnion(
           RegexFactory.createUnion(
                CONCAT_AB,
                CONCAT_AC
           ),
            CONCAT_BC
       ),
       CONCAT_CD
   )
```

**Comment by ignacy123** [comment]
@JHodorHodor I've rewritten some of the more confusing tests, as you rightly pointed out that the readability was lacking :disappointed:
I needed some time to come up with a good way of describing the normalization process through unit tests, but I think it's noticeably better now.

File: app/src/test/kotlin/compiler/lexer/regex/RegexFactoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+private val EMPTY = RegexFactory.createEmpty()
+private val EPSILON = RegexFactory.createEpsilon()
+
+class RegexFactoryTest {
+    private fun <T> assertAllEqual(elements: List<T>) {
+        if (elements.size == 0) return
+        elements.forEach { assertEquals(it, elements[0]) }
+    }
+
+    @Test fun `concat preserves order`() {
```
Comment (JHodorHodor) [inline app/src/test/kotlin/compiler/lexer/regex/RegexFactoryTest.kt:23]: In the other PRs we start naming test with a word "test', so to be consistent, we might do it there also.

**Review by JHodorHodor — APPROVED** [review]
Zostawiłem jeden komentarz dotyczący nazw testów. Poza tym wszystko wygląda dobrze. Testy są dużo czytelniejsze, good job!

**Pull request merged**